Add @count util to torch, use it to track benchmark stats

voznesenskym opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #93013

<img width="1333" alt="image" src="https://user-images.githubusercontent.com/4755252/214687911-f766f072-c162-4298-9aed-c889f1375336.png">

cc @mlazos @soumith @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @chunyuan-w @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire

voznesenskym committed:
wip stats

lint

[ghstack-poisoned]

voznesenskym committed:
Update on "wip stats"


lint

cc mlazos soumith yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

voznesenskym committed:
Update on "wip stats"


lint

cc mlazos soumith yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

voznesenskym committed:
Update on "wip stats"


lint

cc mlazos soumith yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

voznesenskym committed:
Update on "Add @count util to torch, use it to track benchmark stats"


lint

cc mlazos soumith yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

ezyang commented:
This seems fine but:

1. It is duplicative with the existing compiler counters in Dynamo, and
2. The context manager is needlessly costly when you can do a less elegant direct increment call inside the body of the function you want to count, which will only be a bytecode

voznesenskym commented:
> This seems fine but:
> 
> 1. It is duplicative with the existing compiler counters in Dynamo, and

These cant be trivially imported without a bunch of import shuffling, and there is no simple counter here as I could see

> 2. The context manager is needlessly costly when you can do a less elegant direct increment call inside the body of the function you want to count, which will only be a bytecode

This is not a context manager, just a decorator, its functionally equivalent to doing it within the body of the function, right?

ezyang commented on benchmarks/dynamo/common.py:
```diff
@@ -1518,9 +1518,14 @@ def run_one_model(
             print_time_report()
+            stats = f"STATS: call_* op count: {op_count}"
+            for key, value in simple_call_counter.items():
+                stats = f"{stats} | {key}:{value}"
```
nit: `" | ".join(f"{key}:{value}" for key, value in simple_call_counter.items())`, mostly to get rid of the N^2 reinterpolation and it's about the same length of code nayway

voznesenskym replied:
nice

ezyang replied:
I know it doesn't matter, I just reflexively hate n^2 code

ezyang commented on torch/utils/_stats.py:
```diff
@@ -0,0 +1,16 @@
+    @functools.wraps(fn)
+    def wrapper(*args, **kwargs):
+        if fn.__qualname__ not in simple_call_counter:
+            simple_call_counter[fn.__qualname__] = 0
```
One downside to the QOL fix of avoiding having to type qualname is it makes it less easy to grep the codebase to answer the question "given this stat, what line of code am I actually measuring hits for"

voznesenskym committed:
Update on "Add @count util to torch, use it to track benchmark stats"


<img width="1333" alt="image" src="https://user-images.githubusercontent.com/4755252/214687911-f766f072-c162-4298-9aed-c889f1375336.png">

cc mlazos soumith yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

voznesenskym commented:
@pytorchbot merge
